operator/v1alpha1/types_kueue.go

```diff
@@ -0,0 +1,304 @@
+package v1alpha1
+
+import (
+	operatorv1 "github.com/openshift/api/operator/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Kueue is the CRD to represent the kueue operator
+// This CRD defines the configuration that the Kueue
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=kueue,scope=Cluster
+// +k8s:openapi-gen=true
+// +genclient
+// +genclient:nonNamespaced
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'cluster'",message="kueue is a singleton, .metadata.name must be 'cluster'"
+type Kueue struct {
+	metav1.TypeMeta `json:",inline"`
+	// metadata for kueue
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// kueue configuration must not be changed once the object exists
+	// to change the configuration, one can delete the object and create a new object.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="values are immutable once set"
```

Comment:
This is an unusual option on a configuration API for an operator, what's the technical reason for doing this?

Reply:
I put the comment above trying to explain it.

Kueue does not react to config changes. The config are read as the controller starts. So any config change must require a restart of all the pods of the deployment. Essentially we would require a detect of a patch and restart the pods of the deployment. This could would impact HA as we would want to perform an outage of the service to rollout a new config change.

Rather than code that logic, I figured making them immutable is easier.

Reply:
Removing the config, means deleting the deployment and starting again? Isn't that more disruptive?

When the config here changes, you want to update the configmap that is output from this right? Or are you generating the config as an init container?

Either way, you could use either the generation of the CM, or Kueue config, as a way to trigger a rollout of the deployment when the config changes by labelling the podTemplateSpec with a label that refers to the configuration generation

Would simplify the workflow for making updates no?

Reply:
To me, I want the more disruptive change for this. Changing the configuration of Kueue means that we need a complete redloyment of every resource that Kueue created.

If I change integrations from pods to jobs, I need to remove all webhooks reference this.

For upstream kueue, the only option to change configurations is the method I suggested.

The config is mounted as a volume to the main pod. But it is only read before the kueue controllers start.

So a new configuration change would trigger a resync and generate a new config-map but I still need to do a new rollout of the deployment.

Reply:
No, we don't have that code in our operator.

I want to try and minimize the scope of changes for TP as there is so much here.

Trying to guard all these use cases seems much more difficult.

What happens if a admin deletes a webhook, removes a certificate, deletes the deployment of kueue etc.

There are a lot of failure cases here.

Reply:
I'd rather get an actual use case from a customer and implement it rather than trying to do everything at once.

So I thought making it immutable would make some cases easier. We added code to make sure if a cluster CR is deleted it delets all the underlying objects minus the kueue CRs.

And then a user can create a new cluster and it triggers the creation of the new objects based on the Kueue config map.

Reply:
> What happens if a admin deletes a webhook, removes a certificate, deletes the deployment of kueue etc.

Generally operators just put these back if an admin were to remove them, that's part of the operator role, installing and ensuring the resources required to make the operand work

Reply:
Does it have to be TP? Or can it be something post?

Reply:
So you think we should have the ability to patch resources?

I'll remove the immutable requirement and look to support this.

Reply:
Yeah, I think this is fairly fundamental operator behaviour, you could say it's going to be done later, but then I'd ask for a TODO on removing the immutability

Look into adding a label on the deployment podTemplateSpec that represents the config generation and that should trigger the redeploy for you

Reply:
I’ll pull this in I removed that restriction.